packages/gatsby-plugin-sharp/src/process-file.js

```diff
@@ -76,6 +76,13 @@ exports.processFile = (file, transforms, options = {}) => {
   return transforms.map(async transform => {
     try {
       const { outputPath, args } = transform
+
+      // If the outputPath already exists, we created the file at some
+      // point in the past so let's just return.
+      if (fs.existsSync(outputPath)) {
+        return transform
+      }
```

Comment:
Two things:
1. This is quite late - it's good that it's here, but we should not even create transform if file is there already ( and if no trasnform is created then we shouldn't even create a job for it ). What putting this just here would mean is that our "Generating images" progress bar can't be trusted because it would (in case image file already exist) lie that it generates images when in fact it doesn't.
2. In case there is a bug in `gatsby-plugin-sharp` that cause it to generate wrong images - updating to new version (which would contain fix) won't re-create the image because bad image already exists. Before we go forward with this - we need to know if/how we can invalidate/discard previously generated images if they were created with version we found out to be buggy. 
One way to go about (that might not be super complex) is to use some kind of seed or cache key when we generate short hash for the filename of generated images (so when we fix a bug, we change/bump that to ensure hash would change). Need some double checking if this is called in all of our functions (fluid/fixed/resize) - but possibly: https://github.com/gatsbyjs/gatsby/blob/11a0a1642c699a14dc0e8747d03e2a42fe70a679/packages/gatsby-plugin-sharp/src/process-file.js#L227-L231 could handle this. The question would be how can we ensure that we remember about bumping it (manual setting of something like that is very error prone and very easy to miss if we don't have something that can automatically assert / let us know we need to bump it). Using something like `gatsby-plugin-sharp` package version would work automatically but at expense of invalidating more often than it actually needs to.

Reply:
re) 1 👍  — my PRs follow the make it work, make it right, make it fast trajectory ;-) so I'll go make it right now.

re) 2 — yeah I've been playing around with some ideas for a new jobs API and one of them is that job creators _have_ to add a cache version. I'm strongly against tying the cache to the version number as that changes all the time and would make the cache somewhat pointless. I can the chance of some maintainer forgetting to bump the cache key but the worst case scenario is it'd take another or two for someone to notice that the buggy output was still buggy and then the cache key gets incremented — this sounds much nicer than constantly having to rebuild job outputs.

Reply:
Agree with not tying cache key to plugin version (I just listed this as possibility) - it has obvious issues with being overly aggressive with clearing things for builds when in reality we very rarely (if ever) will need to change it. On top of changing file names that impact not only builds but also visitors of the sites (image that didn't in reality changing url needlesly).

Reply:
For the potential automatic part to discover when we need to bump cache key and not rely only on developers remembering about it (quite error prone): @ascorbic as part of his work on revamped `gatsby-image` added visual-regressions tests ( https://github.com/gatsbyjs/gatsby/blob/master/e2e-tests/visual-regression/cypress/integration/image.js ) that does image diff comparison for what is displayed in the browser against snapshoted/fixture screenshots ( https://github.com/gatsbyjs/gatsby/tree/master/e2e-tests/visual-regression/cypress/snapshots/image.js ). He did mentioned that those tests already proved useful couple of time when it caught regressions (or at least changes to what is snapshoted/fixture vs what is displayed currently) before PRs were even merged.

Those tests can potentially incidentaly discover issues / changes in what is being generated by `gatsby-plugin-sharp`, but it's not their goal and they also focus on various scenarios specific to `gatsby-image` (so things like various viewports or GatsbyImage/StaticImage scenarios and not really about testing if `gatsby-plugin-sharp` generates correct things).

We could use something similar for `gatsby-plugin-sharp` tests that we would prepare transform objects "matrix" that runs `processFile` thingie for multiple combinations of args https://github.com/gatsbyjs/gatsby/blob/11a0a1642c699a14dc0e8747d03e2a42fe70a679/packages/gatsby-plugin-sharp/src/process-file.js#L33-L48

so get few source images of different formats (jpg,png,etc), have various width, height, quality, toFormat etc and generate them all and to diff comparison to baseline/fixtures/snapshots which could tell us when something changes and if something changes then we figure out if code change is buggy or if we need to bump cache key (and update snapshots) because change is actually expected

Reply:
I think the visual diff would be a really useful addition to the image tests. It could be simpler than the visual regression tests, as it wouldn't need all the cypress handling, and could just use jest image snapshot directly in unit tests

Reply:
Yup those, shouldn't be e2e - we should compare files - possibly even check file signatures or do some image comparison ( not sure if we would need image file diff or not - depends if we want/need some diff threshold or exact comparison would work - things like cross platform changes etc might require us to do image diff vs file diff)

Reply:
and also just to clarify - I'm not saying we need to do all of above in this PR. We aren't changing how images are generated here. What I wanted is mainly plan for
> Before we go forward with this - we need to know if/how we can invalidate/discard previously generated images if they were created with version we found out to be buggy.

and not necesairly implement this right away. So even https://github.com/gatsbyjs/gatsby/pull/28330/commits/8e55552b8e179a40e7b3464fbc9d52a22d48e722 doesn't need to be here ATM

Reply:
on the cache key... I was just thinking that while we're talking about not changing generated URLs unless it really needs to happen — I had just added a change to every URL 🤦 

So removed that in https://github.com/gatsbyjs/gatsby/pull/28330/commits/48c4a21fa53a6020bbe19b8b35bc0554e911e48c — let's reserve that for an actual bug fix.

Also 👍 to jest image snapshot tests.